File: M3 Intermediate DSA/src/L6_Arrays_Carry_Forward/Assignment2.java
package L6_Arrays_Carry_Forward;

import java.util.ArrayList;

public class Assignment2 {

  public int[] solve(int[] A) {
    int N = A.length;
    int maxValue = A[N-1];
    ArrayList<Integer> ans = new ArrayList<>();
    ans.add(A[N-1]);

    for (int index = N -2; index >= 0; index--) {

      if (A[index] > maxValue) {
        maxValue = A[index];
        ans.add(A[index]);
      }

    }

    return ans.stream().mapToInt(i -> i).toArray();
  }
}

File: M3_Intermediate_DSA/L4_Carry_Forward/assignment2.go
package main

func SolveA2(A []int) []int {
	N := len(A)
	maxValue := A[N-1]
	ans := []int{maxValue}

	for index := N - 2; index >= 0; index-- {
		if A[index] > maxValue {
			maxValue = A[index]
			ans = append(ans, A[index])
		}
	}

	return ans
}
